src/pinnable_mapped_file.cpp

```diff
@@ -227,54 +229,63 @@ void pinnable_mapped_file::setup_copy_on_write_mapping() {
    // non-sharable chainbase dbs using mapped mode are flushed to disk
    // ----------------------------------------------------------------------------------
    for (auto pmm : _instance_tracker)
-      pmm->save_database_file(true);
+      pmm->save_database_file(true, false);
 
-   _file_mapped_region = bip::mapped_region(_file_mapping, bip::copy_on_write);
-   *((char*)_file_mapped_region.get_address()+header_dirty_bit_offset) = dirty; // set dirty bit in our memory mapping
-
-   _segment_manager = reinterpret_cast<segment_manager*>((char*)_file_mapped_region.get_address()+header_size);
+   _cow_address = mmap(NULL, _database_size, PROT_READ | PROT_WRITE, MAP_PRIVATE, _file_mapping.get_mapping_handle().handle, 0);
```

Comment:
I've noticed a couple of issues here when testing,

The size should be the size of the DB, not the configured size. These will differ when a user tries to shrink the DB (not supported). Just looking through the code some I'm worried we might have broke this in other ways in #21 (such as when writing out heap mode), and may require a separate fix for 5.0. I'm also curious why grow_shrink test isn't failing; maybe the test needs to be improved, unsure.

If this fails, nodeos crashes leaving the DB dirty. This is fairly easy for someone to stumble on: if I only have a box with 24GB of RAM and configure `chain-state-db-size-mb = 24576`, it will fail. Need to be checking the return value here and more gracefully fail.

Reply:
> The size should be the size of the DB, not the configured size. 

Hum, I'm not following. Here we use `_database_size`, which is the `shared_file_size` passed to the `database` and `pinnable_mapped_file` constructors.

Reply:
> I'm worried we might have broke this in other ways in https://github.com/AntelopeIO/chainbase/pull/21 (such as when writing out heap mode)

Do you have any specific issue in mind. That is the main reason I added the `_database_size` member, to ensure we stored the requested database size, so we could precisely write this size, excluding extra space required for huge pages alignment or such.

Reply:
The requested database size isn't the actual database size in case of,
https://github.com/AntelopeIO/chainbase/blob/7615ddab287e06fd31f800e66fe39b3a19320ec8/src/pinnable_mapped_file.cpp#L136-L140
So, for example, if you first `chain-state-db-size-mb = 24576` and then restart with `chain-state-db-size-mb = 1024` the DB size will remain 24576 but `_database_size` is 1024 and it results in a crash with `mapped_private`

It also probably results in heap mode not being completely written out now too for similar reason: `_database_size` is requested size, not actual size. Basically any place we use `_database_size` is suspect currently.

Reply:
Thanks! I fixed the issue in this PR, and if you agree with the change, I'll create a PR for fixing it in `5.0`